tests/Control/UserDefinedFormControllerTest.php

```diff
@@ -384,6 +397,437 @@ public function testRecipientSubjectMergeFields()
+
+        $this->logInWithPermission('CMS_ACCESS_CMSMain');
+        $response = $this->post($url, ['ID' => $fieldID]);
+        $this->assertEquals(
+            403,
+            $response->getStatusCode(),
+            'Users without CMS_ACCESS_AssetAdmin can\'t confirm folder'
+        );
+
+        $this->logInWithPermission(['CMS_ACCESS_CMSMain', 'CMS_ACCESS_AssetAdmin']);
+        $response = $this->post($url, ['ID' => $fieldID]);
+        $this->assertEquals(200, $response->getStatusCode(), 'CMS editors can access confirm folder form ');
+
+        $response = $this->post($url, ['ID' => $restrictedFieldID]);
+        $this->assertEquals(
+            403,
+            $response->getStatusCode(),
+            'CMS editors can\'t confirm folder form for restricted form'
+        );
+
+        $this->logInWithPermission('ADMIN');
+
+        $response = $this->post($url, ['ID' => $restrictedFieldID]);
+        $this->assertEquals(
+            200,
+            $response->getStatusCode(),
+            'Admins can confirm folder form for restricted form'
+        );
+    }
+
+    public function testConfirmfolderExistingFolder()
+    {
+        $this->logInWithPermission(['CMS_ACCESS_CMSMain', 'CMS_ACCESS_AssetAdmin']);
+
+        $url = 'UserDefinedFormController/confirmfolder?';
+        $fieldID = $this->idFromFixture(EditableFileField::class, 'file-field-1');
+        $folderID = $this->idFromFixture(Folder::class, 'restricted');
+
+        $response = $this->post($url, ['ID' => $fieldID, 'FolderOptions' => 'existing', 'FolderID' => $folderID]);
+        $this->assertEquals(200, $response->getStatusCode(), 'Valid request to confirm an existing folder is successful');
+        $this->assertEquals(
+            $folderID,
+            EditableFileField::get()->byID($fieldID)->FolderID,
+            'FileField points to restricted folder'
+        );
+    }
+
+    public function testConfirmfolderInexistingFolder()
+    {
+        $this->logInWithPermission(['CMS_ACCESS_CMSMain', 'CMS_ACCESS_AssetAdmin']);
+
+        $url = 'UserDefinedFormController/confirmfolder?';
+        $fieldID = $this->idFromFixture(EditableFileField::class, 'file-field-1');
+
+        $response = $this->post($url, ['ID' => $fieldID, 'FolderOptions' => 'existing', 'FolderID' => -1]);
+        $this->assertEquals(400, $response->getStatusCode(), 'Confirm a non-existant folder fails with 400');
+    }
+
+    public function testConfirmfolderRootFolder()
+    {
+        $this->logInWithPermission(['CMS_ACCESS_CMSMain', 'CMS_ACCESS_AssetAdmin']);
+
+        $url = 'UserDefinedFormController/confirmfolder?';
+        $fieldID = $this->idFromFixture(EditableFileField::class, 'file-field-1');
+
+        $response = $this->post($url, ['ID' => $fieldID, 'FolderOptions' => 'existing', 'FolderID' => 0]);
+        $this->assertEquals(200, $response->getStatusCode(), 'Valid request to confirm an root folder is successful');
```

Comment:
This test is failing because we don't handle the case where the user decides to set the root folder as the the target for the file field.